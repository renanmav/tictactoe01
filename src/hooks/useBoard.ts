import { useEffect, useState } from 'react';

import { Difficulties } from '../screens/SelectDifficulty';

import { useWinner } from './useWinner';

// @ts-ignore
String.prototype.replaceAt = function (index: number, replacement: string) {
  return this.substr(0, index) + replacement + this.substr(index + replacement.length);
};

export function useBoard(difficulty: Difficulties) {
  const [board, setBoard] = useState('         ');
  const [isMyTurn, setIsMyTurn] = useState(true);

  const { winner, isEvaluating, winnerMatch, reset } = useWinner(board);

  useEffect(() => {
    let isCurrent = true;

    if (isMyTurn || winner !== '?' || isEvaluating) {
      return;
    }

    (async () => {
      const body = {
        board,
        level: difficulty,
      };
      const resp = await fetch('https://tictactoe.api.01card.com.br', {
        method: 'POST',
        body: JSON.stringify(body),
        headers: {
          authorization: 'Bearer ck9tzriy6000201l19kdy5d4i',
        },
      });
      const data = await resp.json();
      if (isCurrent) {
        setIsMyTurn(true);
        setBoard(data.board);
      }
    })();

    return () => {
      isCurrent = false;
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [winner, board]);

  function updateBoard(index: number, symbol: 'X' | 'O') {
    if (!isMyTurn || winner !== '?' || board.charAt(index) !== ' ') {
      return;
    }
    // @ts-ignore
    const newBoard = board.replaceAt(index, symbol);
    setBoard(newBoard);
    setIsMyTurn(false);
  }

  function restart() {
    setIsMyTurn(true);
    setBoard('         ');
    reset();
  }

  return { board, updateBoard, winner, match: winnerMatch, restart };
}
