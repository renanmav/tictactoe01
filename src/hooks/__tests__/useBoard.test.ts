import { wait } from '@testing-library/react-native';
import { renderHook, act } from '@testing-library/react-hooks';
import { FetchMock } from 'jest-fetch-mock';

import { useBoard } from '../useBoard';

beforeEach(() => {
  (fetch as FetchMock).resetMocks();
});

describe('useBoard', () => {
  test('if board, updateBoard and winner are returned', () => {
    const { result } = renderHook(() => useBoard('easy'));

    expect(result.current.board).toBe('         ');
    expect(result.current.updateBoard).toBeDefined();
    expect(result.current.winner).toBe('?');
    expect(fetch).not.toBeCalled();
  });

  test('if board is updated and it is my turn, a new board is returned', async () => {
    (fetch as FetchMock).mockResponseOnce(JSON.stringify({ board: 'XO       ' }));

    const { result } = renderHook(() => useBoard('easy'));

    expect(result.current.updateBoard).toBeDefined();

    act(() => result.current.updateBoard(0, 'X'));

    await wait(() => {
      expect(result.current.board).toBe('XO       ');
      expect(result.current.winner).toBe('?');
      expect(fetch).toBeCalledTimes(1);
      expect((fetch as FetchMock).mock.calls).toMatchSnapshot();
    });
  });

  test('if there is a winner, it should not fetch API', async () => {
    const botMoves = ['XO       ', 'XOO X    '];

    (fetch as FetchMock)
      .mockResponseOnce(JSON.stringify({ board: botMoves[0] }))
      .mockResponseOnce(JSON.stringify({ board: botMoves[1] }));

    const { result } = renderHook(() => useBoard('easy'));

    expect(result.current.updateBoard).toBeDefined();

    act(() => result.current.updateBoard(0, 'X'));
    await wait(() => expect(result.current.board).toBe(botMoves[0]));

    act(() => result.current.updateBoard(4, 'X'));
    await wait(() => expect(result.current.board).toBe(botMoves[1]));

    act(() => result.current.updateBoard(8, 'X')); // X wins

    await wait(() => {
      expect(result.current.winner).toBe('X');
      expect(result.current.board).toBe('XOO X   X');
      expect(fetch).toBeCalledTimes(3);
      expect((fetch as FetchMock).mock.calls).toMatchSnapshot();
    });
  });
});
