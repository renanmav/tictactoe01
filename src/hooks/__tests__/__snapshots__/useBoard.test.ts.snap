// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useBoard if all fields are returned 1`] = `Array []`;

exports[`useBoard if board is updated and it is my turn, a new board is returned 1`] = `
Array [
  Array [
    "https://tictactoe.api.01card.com.br",
    Object {
      "body": "{\\"board\\":\\"X        \\",\\"level\\":\\"easy\\"}",
      "headers": Object {
        "authorization": "Bearer ck9tzriy6000201l19kdy5d4i",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`useBoard if board is updated and it is my turn, a new board is returned 2`] = `Array []`;

exports[`useBoard if there is a winner, it should not fetch API 1`] = `
Array [
  Array [
    "https://tictactoe.api.01card.com.br",
    Object {
      "body": "{\\"board\\":\\"X        \\",\\"level\\":\\"easy\\"}",
      "headers": Object {
        "authorization": "Bearer ck9tzriy6000201l19kdy5d4i",
      },
      "method": "POST",
    },
  ],
  Array [
    "https://tictactoe.api.01card.com.br",
    Object {
      "body": "{\\"board\\":\\"XO  X    \\",\\"level\\":\\"easy\\"}",
      "headers": Object {
        "authorization": "Bearer ck9tzriy6000201l19kdy5d4i",
      },
      "method": "POST",
    },
  ],
  Array [
    "https://tictactoe.api.01card.com.br",
    Object {
      "body": "{\\"board\\":\\"XOO X   X\\",\\"level\\":\\"easy\\"}",
      "headers": Object {
        "authorization": "Bearer ck9tzriy6000201l19kdy5d4i",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`useBoard if there is a winner, it should not fetch API 2`] = `
Array [
  0,
  4,
  8,
]
`;
